### PR TITLE
[feat] Made the Backend of Template Creation

### PR DESCRIPTION
**Pull request by CodyGarciaa**
### 🦋 Description
[//]: # "Required - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."
- holy balls this sprint was a lot lol (but it was also technically 2 sprints combined so fire)
- created all the flowy stuff of template, like non-pretty stuff
- created reusable template functions to edit template stuff from supabase (creating, editing, and deleting)
- created a kinda form layout that hella like nested objects (start from parent, go to child, and child; so theres like a grandchild component?)
- as the user is filling out template creation, everything is getting saved locally until the user submits the template, then everything gets pushed to supabase all at once

### 🦧 Screenshots
[//]: # "Required for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior 
- new template
 new template
- new roles
new roles
- remove roles
remove roles
- new phases
new phases
- new prompts
new prompts
- remove phases and prompts
remove phases prompts
- submit template (shouldnt take this long, i was just doing in the car on hotspot lol
submit template

## 🐸 How to review
[//]: # 'Required - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?'
- so first get familiar with some stuff, kinda glance at least a bit at `schema.d.ts` and `api/supabase/queries/templates.tsx`. these are the types i either used, changed, or created to keep at the typescript shit happy. and then the `templates.tsx` stuff are the general functions i used to add to supabase. overall just know each adds a row to supabase + what they need to take in.
- here comes the main shit... `app/templates/page.tsx` `TemplateBuilder.tsx` `RoleForm.tsx` and `TemplateOverviewForm.tsx`
- its kinda linear the files: `app/templates/page.tsx` -> `TemplateBuilder.tsx` -> `RoleForm.tsx` & `TemplateOverviewForm.tsx`

- so it starts at `app/templates/page.tsx`. pretty small file, this file holds the title of the page, the routing `http://localhost:3000/templates` routes to this `page.tsx` technically, and then the button to start creating a template -> `TemplateBuilder`. The reason I separated page.tsx from TemplateBuilder bc its allows it to be easier to merge this feature with the rest of the app if needed (like just need to change the button/what calls the button to `TemplateBuilder`).
Screenshot 2025-10-30 at 9 10 39 PM
- the only other main thing to note from `page.tsx` is that this is where the `LocalStore` is created. **LOCALSTORE IS CRUTIAL**. basically, `LocalStore` is this semi-involved object that temporarily stores the form data until pushed to supabase. it moves around hella bc it gets passed between different components bc everything saves to it. 
Screenshot 2025-10-30 at 9 15 34 PM

- now to `TemplateBuilder.tsx`. Highkey gonna be the most confusing file (271 lines of code as of now). this is the one imma try to condense mainly after submitting the pr. so don't be surprised if it gets broken down to more but smaller files in the future. hella subfunctions there.
- `useForceUpdate()` is just a random useState to force a rerender of the page when needed.
- `addRole()`, `removeRole()`, `renameRole()` (lowk made earlier on, i think its irrelevant now tho tbh, might remove later), `addPhase()`, `removePhase()`, `addPrompt()`, and `removePrompt()` all do that it sounds like.
- `setActiveUpdate()` is a function called so that changes in the children components `RoleForm.tsx` and `TemplateOverviewForm.tsx` gets made in this file.
- `saveTemplate()` is a function called at the end that just pushed all the stuff from this new template into supabase. FIRE

- `TemplateOverviewForm.tsx` is a pretty simple file as well. all its in charge of is the "Scenario Overview" tab and stuff there. it takes in the `Template` object to READ what the template has. Note, doesn't do any WRITING to the `Template` object here. all changes gets sent to the `onChange` function that gets passed in (`setActiveUpdate` in `TemplateBuilder.tsx`

- `RoleForm.tsx` same vibes as `TemplateOverviewForm.tsx`. might be a little more involved because there's just more information getting passed back and forth, but overall same logic.
 

## 🐙 Next steps
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."
- lowk the supabase testing code doesnt work anymore lolll. before using the functions in the frontend code it worked, but then when i actually started using the functions i refractored them hella so that code is outdated. dont look at lol
- code is kinda messy rn sorrrrry. after submitting this pr ill try to make more readable and/or with more comments, but i wanted to have the pr in first at least (while code works fr)
- saving doesnt happen incrementally save (so like if u refresh mid-editing it'll reset rip). save only happens once submitted.

### Online sources
[//]: # 'Copy links to any tutorials or documentation that was useful to you when working on this PR'
- again chat came in clutch (my structure for everything was getting confusing and it helped to tell me best ways to hold everything so that it works but also not terribly inefficient)

### 🐁 Related PRs
[//]: # "Add related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

CC: @me-liu

### CODE REVIEW

File: app/templates/components/TemplateBuilder.tsx [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+    // call whenever my screen isn't updating :)
+    setTick(tick => (tick + 1) % 10);
+  }
+
+  function createUUID(): UUID {
+    //helper function to create new UUIDs
+    return crypto.randomUUID() as `${string}-${string}-${string}-${string}-${string}`;
+  }
+
+  function addRole(): void {
```
Comment (me-liu) [inline app/templates/components/TemplateBuilder.tsx]: for ex this is what addRole would look like w the new package:
```
function addRole(): void {
  if (!localStore) return;
  const newRoleID = createUUID();

  updateLocalStore(draft => {
    draft.rolesById[newRoleID] = {
      role_id: newRoleID,
      role_name: "New Role",
      role_description: "New Role Description",
      template_id: draft.templateID,
    };

    draft.roleIds.push(newRoleID);
    draft.rolePhaseIndex[newRoleID] = {};

    for (const phaseID of draft.phaseIds) {
      const newRolePhaseID = createUUID();
      draft.rolePhaseIndex[newRoleID][phaseID] = newRolePhaseID;
      draft.rolePhasesById[newRolePhaseID] = {
        role_phase_id: newRolePhaseID,
        phase_id: phaseID,
        role_id: newRoleID,
        description: null,
      };
      draft.promptIndex[newRolePhaseID] = [];
    }
  });

  setActiveId(newRoleID);
}
